lib/spack/spack/compilers/msvc.py

```diff
@@ -231,24 +231,45 @@ def msvc_version(self):
 
     @property
     def short_msvc_version(self):
+        """This is the shorthand VCToolset version of form
+        MSVC<short-ver>
         """
-        This is the shorthand VCToolset version of form
-        MSVC<short-ver> *NOT* the full version, for that see
+        return "MSVC" + self.vc_toolset_ver
+
+    @property
+    def vc_toolset_ver(self):
+        """
+        The toolset version is the version of the combined set of cl and link
+        This typically relates directly to VS version i.e. VS 2022 is v143
+        VS 19 is v142, etc.
+
+        This is *NOT* the full version, for that see
         Msvc.msvc_version or MSVC.platform_toolset_ver for the
         raw platform toolset version
+
         """
-        ver = self.platform_toolset_ver
-        return "MSVC" + ver
+        ver = self.msvc_version[:2].joined.string[:3]
+        return ver
 
     @property
     def platform_toolset_ver(self):
         """
         This is the platform toolset version of current MSVC compiler
-        i.e. 142.
+        i.e. 142. The platform toolset is the targeted MSVC library/compiler
+        versions by compilation (this is different from the VC Toolset)
+
+
         This is different from the VC toolset version as established
-        by `short_msvc_version`
+        by `short_msvc_version`, but typically are represented by the same
+        three digit value
         """
-        return self.msvc_version[:2].joined.string[:3]
+        # VS22 introduces the first divergence of VS toolset version
+        # (144 for "recent" releases) and platform toolset version (143)
+        # so it needs additional handling until MS releases v144
+        # or adds better support for detection
```

Comment:
> I think it might make sense to add a TODO, given that this approach is likely to "expire"
> 
> Also IMO it's clearest if a comment outlining an exception starts by outlining the normal behavior like
> 
> ```
> # Typically VS toolset version and platform toolset versions match
> ```

Makes sense, will do.

> Also, I would have expected the vs22 check to be
> 
> ```
> vs22_toolset = Version(toolset_ver) > Version("143")
> ```

These are functionally the same. however for VS 22, the Platform Toolset is v143 and the VC toolset ranges depending on your VS version from 143 to 144. 142 is VS 17 for both VC Toolset and Platform Toolset. I wanted to express "is vs22" by expressing "version is newer than VS 17 version".

> but did the vs toolset version skip 143 altogether? If so, I would recommend the following comment:
> 

It did not.

> 
> We could also potentially make this more robust by speculatively restricting the range to only return 143 for vs toolset versions < 145 (i.e. return to matching vs/platform toolsets for later VS versions).

I'd prefer to avoid this particular constraint in favor of the TODO you mention above, as this to me would get more confusing as MS increments versions and the VC toolset and platform toolset may or may not go in and out of sync. I think the TODO should hold us until MS pushes an update, and in that case we can decide what to do (they may not go back to vc toolset == platform toolset).

Reply:
> the VC toolset ranges depending on your VS version from 143 to 144. 

Ah ok so that detail in some form would be useful in the comment (i.e. "there was a vc toolset update without an accompanying platform toolset version bump")

> I'd prefer to avoid this particular constraint in favor of the TODO you mention above, as this to me would get more confusing as MS increments versions and the VC toolset and platform toolset may or may not go in and out of sync.

That's fair. It's somewhat idiomatic for our packages to attempt to address strange per-version behavior just for that version and revisit, but not so established I wanted to mandate that here (the TODO should be sufficient).

Reply:
> Ah ok so that detail in some form would be useful in the comment (i.e. "there was a vc toolset update without an accompanying platform toolset version bump")

So technically the entire MS community (Cmake, XMake, etc) has been relying on this being true, but nowhere in the MS docs does it actually say it's supposed to be this way. It's more that MS has changed the versioning sentiment for the VC toolset (previously they were only updating the rev/build numbers, not the minor version). A blog for more context: https://devblogs.microsoft.com/cppblog/msvc-toolset-minor-version-number-14-40-in-vs-2022-v17-10/

Happily updating the doc to reflect this additional context, just thought I'd give you a bit more context so you're not confused when my update is slightly different from the request

Reply:
(also the blog explains the implications of the update better than I am)